File: pascal'sTriangle.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :NO
// Your code here along with comments explaining your approach
/*
 * We calulate values in every row using the previous row and  j-1 and jth term to calculate the value.
 * we need to create a temporary list for every row and after completion of evry row add it to the result list.
*/
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            List<Integer> list = new ArrayList<>();
            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    list.add(1);
                } else {
                    int left = result.get(i - 1).get(j - 1);
                    int right = result.get(i - 1).get(j);

                    list.add(left + right);
                }
            }

            result.add(list);
        }

        return result;
    }
}

File: pascal's triangle.py
# // Time Complexity : O(N)
# // Space Complexity : O(N)
# // Did this code successfully run on Leetcode : YES
# // Any problem you faced while coding this : NO


# // Your code here along with comments explaining your approach
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 1:
            return [[1]]
            
        resList = []
        resList.append([1])
        resList.append([1, 1])

        for i in range(2, numRows):
            currList = [1 for _ in range(i+1)]
            prevListLen = len(resList[i-1])
            for j in range(1, prevListLen):
                left = resList[i-1][j-1]
                right = resList[i-1][j]

                currList[j] = left + right
            resList.append(currList)  
        return resList       
